keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,655 @@
++ _greenfield bucket_ - a new bucket created by automation.
++ _object_ - an atomic, immutable unit of data stored in buckets.
++ _provisioner_ - a generic term meant to describe the combination of a sidecar and driver. "Provisioning" a bucket can mean creating a new bucket or granting access to an existing bucket.
++ _sidecar_ - a Kubernetes-aware container (usually paired with a driver) that is reponsible for fullfilling COSI requests with the driver via gRPC calls. The sidecar's access can be restricted to the namespace where it resides, which is expected to be the same namespace as the provisioner. The COSI sidecar is provided by the Kubernetes community.
++ _storage instance_ - refers to the back object storage endpoint being abstracted by the Bucket API (a.k.a “bucket” or “container”).
++ _driverless_ - a system where no driver is deployed to automate object store operations.  COSI automation may still be deployed to managed COSI APIs. **Note:** the current state of the KEP does not justify nor define driverless aka "static provisioning". If the community feels this is an MVP requirement we will need to flush this out.
+
+# Proposal
+
+## User Stories
+
+#### Admin
+
+- As a cluster administrator, I can control access to new and existing buckets when accessed from the cluster, regardless of the backing object store.
+
+#### User
+
+- As a developer, I can define my object storage needs in the same manifest as my workload, so that deployments are streamlined and encapsulated within the Kubernetes interface.
+- As a developer, I can define a manifest containing my workload and object storage configuration once, so that my app may be ported between clusters as long as the storage provided supports my designated data path protocol.
+- As a developer, I want to create a workload controller which is bucket API aware, so that it can dynamically connect workloads to object storage instances.
+
+## APIs
+
+### Storage APIs
+
+#### BucketRequest
+
+A user facing, namespaced custom resource requesting a bucket endpoint. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is required in order to grant credentialed access to the bucket. 
+
+**NOTE:** _Currently under review: we are discussing whether or not a BR is needed for any brownfield use-cases. We are considering BRs only apply to greenfield bucket creation (where a BAR is also needed), and that all brownfield accesses need only a BAR._
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  protocol: [3]
+  bucketPrefix: [4]
+  bucketClassName: [5]
+  bucketInstanceName: [6]
+status:
+  phase: [7]
```

Comment:
According to API conventions (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties), `The pattern of using phase is deprecated. Newer API types should use conditions instead. `

Reply:
Thanks @xing-yang !  From this link:
```
The pattern of using phase is deprecated. Newer API types should use conditions instead. Phase was essentially a state-machine enumeration field, that contradicted system-design principles and hampered evolution, since adding new enum values breaks backward compatibility. Rather than encouraging clients to infer implicit properties from phases, we prefer to explicitly expose the individual conditions that clients need to monitor. Conditions also have the benefit that it is possible to create some conditions with uniform meaning across all resource types, while still exposing others that are unique to specific resource types. See #7856 for more details and discussion.
```